pkg/clients/s3/s3.go

```diff
@@ -18,262 +18,96 @@ package s3
-	b := Bucket{}
-	bucketVersioning, err := c.s3.GetBucketVersioningRequest(&s3.GetBucketVersioningInput{Bucket: aws.String(bucket.GetBucketName())}).Send(context.TODO())
-	if err != nil {
-		return nil, err
-	}
-	b.Versioning = bucketVersioning.Status == s3.BucketVersioningStatusEnabled
-	policyVersion, err := c.iamClient.GetPolicyVersion(username)
-	if err != nil {
+func NewClient(ctx context.Context, credentials []byte, region string, auth awsclients.AuthMethod) (Client, error) {
+	cfg, err := auth(ctx, credentials, awsclients.DefaultSection, region)
+	if cfg == nil {
 		return nil, err
 	}
-	b.UserPolicyVersion = policyVersion
-
-	return &b, err
+	return s3.New(*cfg), err
 }
 
-// CreateUser - Create as user to access bucket per permissions in BucketSpec returing access key and policy version
-func (c *Client) CreateUser(username string, bucket *v1alpha3.S3Bucket) (*iam.AccessKey, string, error) {
-	policyDocument, err := newPolicyDocument(bucket)
-	if err != nil {
-		return nil, "", fmt.Errorf("could not update policy, %s", err.Error())
-	}
-	accessKeys, err := c.iamClient.CreateUser(username)
-	if err != nil {
-		return nil, "", fmt.Errorf("could not create user %s", err)
-	}
+// CreatePatch creates a *v1beta1.RDSInstanceParameters that has only the changed
+// values between the target *v1beta1.RDSInstanceParameters and the current
+// *rds.DBInstance
+func CreatePatch(p v1beta1.S3BucketParameters, policy *string) (*v1beta1.S3BucketParameters, error) {
+	patch := &v1beta1.S3BucketParameters{}
 
-	currentVersion, err := c.iamClient.CreatePolicyAndAttach(username, username, policyDocument)
-	if err != nil {
-		return nil, "", fmt.Errorf("could not create policy %s", err)
+	if p.Policy != nil && *p.Policy != *policy {
+		patch.Policy = p.Policy
 	}
 
-	return accessKeys, currentVersion, nil
+	return patch, nil
 }
 
-// UpdateBucketACL - Updated CannedACL on Bucket
-func (c *Client) UpdateBucketACL(bucket *v1alpha3.S3Bucket) error {
-	var err error
-	name := bucket.GetBucketName()
-	if bucket.Spec.CannedACL != nil {
-		input := &s3.PutBucketAclInput{
-			ACL:    *bucket.Spec.CannedACL,
-			Bucket: &name,
-		}
-		_, err = c.s3.PutBucketACLRequest(input).Send(context.TODO())
-	}
-
-	return err
-}
-
-// UpdateVersioning configuration for Bucket
-func (c *Client) UpdateVersioning(bucket *v1alpha3.S3Bucket) error {
-	versioningStatus := s3.BucketVersioningStatusSuspended
-	if bucket.Spec.Versioning {
-		versioningStatus = s3.BucketVersioningStatusEnabled
-	}
-	name := bucket.GetBucketName()
-	input := &s3.PutBucketVersioningInput{Bucket: &name, VersioningConfiguration: &s3.VersioningConfiguration{Status: versioningStatus}}
-	_, err := c.s3.PutBucketVersioningRequest(input).Send(context.TODO())
+// IsUpToDate checks if the current spec is in sync with the observed bucket properties.
+func IsUpToDate(p v1beta1.S3BucketParameters, policy string) (bool, error) {
```

Comment:
Normally, what `IsUpToDate` does is that it checks `*Parameters` struct with actual cloud resource struct. However, in this case we have one-to-many relation between `*Parameters` struct and sub-resources. The easiest solution would be to pass in all those resources as arguments to this function, but this would bloat the function signature. So, I think doing the actual `GET` call of each resource and comparing it against `p v1beta1.S3BucketParameters` would make sense for this case.

Right now, what it'd do is to `GET` the policy and compare it with the policy that `p` has and return the result. I'd not use `CreatePatch` as the logic is simple enough.

Reply:
Right.
I created the `CreatePatch` method considering there will be more properties to added.